decode.go

```diff
@@ -942,36 +942,45 @@ func newMapReader(r reader) (*mapReader, error) {
 		return nil, err
 	}
 
+	// encoding size in bytes uint8 or uint32
+	encSize := 1
+
 	var n uint32
+	var c uint32
 	switch amqpType(b) {
 	case typeCodeNull:
 		return nil, errNull
 	case typeCodeMap8:
-		bn, err := r.ReadByte()
+		b, err = r.ReadByte()
 		if err != nil {
 			return nil, err
 		}
-		n = uint32(bn)
+		n = uint32(b)
+		b, err = r.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+		c = uint32(b)
 	case typeCodeMap32:
 		err = binary.Read(r, binary.BigEndian, &n)
 		if err != nil {
 			return nil, err
 		}
+		err = binary.Read(r, binary.BigEndian, &c)
+		if err != nil {
+			return nil, err
+		}
+		encSize = 4
 	default:
 		return nil, errorErrorf("invalid map type %x", b)
 	}
 
-	if uint64(n) > uint64(r.Len()) {
+	if uint64(n) > uint64(r.Len()+encSize) {
```

Comment:
Is this correct? The `encSize` should have already been read from the `reader` at this point.

Reply:
`r.Len()` is pairs length here, but `n` is expected to be size of `count` parameter plus pairs, e.g: 
```
payload = {0x4, 0x2, 0xa, 0xb}
number = 0x4
count = 0x2
pairs = {0xa, 0xb}

number == sizeof(count) + len(pairs)
```

Reply:
Oh, that makes sense, thank you.